src/report.rs

```diff
@@ -0,0 +1,249 @@
+/// fn main() -> Report<PlaceholderError> {
+///     let _v = may_fail_with_placeholder_error()?;
+///
+///     Report::ok()
+/// }
+/// # #[derive(Debug, Snafu)]
+/// # struct PlaceholderError;
+/// # fn may_fail_with_placeholder_error() -> Result<u8, PlaceholderError> { Ok(42) }
+/// ```
+///
+/// [try-ff]: crate::guide::feature_flags#unstable-try-trait
+///
+/// ## Interaction with the Provider API
+///
+/// If you return a [`Report`][] from your function and enable the
+/// [`unstable-provider-api` feature flag][provider-ff], additional
+/// capabilities will be added:
+///
+/// 1. If provided, a [`Backtrace`][] will be included in the output.
+/// 1. If provided, a [`ExitCode`][] will be used as the return value.
+///
+/// [provider-ff]: crate::guide::feature_flags#unstable-provider-api
+/// [`Backtrace`]: crate::Backtrace
+/// [`ExitCode`]: std::process::ExitCode
+pub struct Report<E>(Result<(), E>);
+
+impl<E> Report<E> {
+    /// Convert an error into a [`Report`][].
+    ///
+    /// Recommended if you support versions of Rust before 1.61.
+    ///
+    /// ```rust
+    /// use snafu::{prelude::*, Report};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// struct PlaceholderError;
+    ///
+    /// fn main() -> Result<(), Report<PlaceholderError>> {
+    ///     let _v = may_fail_with_placeholder_error().map_err(Report::from_error)?;
+    ///     Ok(())
+    /// }
+    ///
+    /// fn may_fail_with_placeholder_error() -> Result<u8, PlaceholderError> {
+    ///     Ok(42)
+    /// }
+    /// ```
+    pub fn from_error(error: E) -> Self {
+        Self(Err(error))
+    }
+
+    /// Executes a closure that returns a [`Result`][], converting the
+    /// error variant into a [`Report`][].
+    ///
+    /// Recommended if you support versions of Rust before 1.61.
+    ///
+    /// ```rust
+    /// use snafu::{prelude::*, Report};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// struct PlaceholderError;
+    ///
+    /// fn main() -> Result<(), Report<PlaceholderError>> {
+    ///     Report::capture_into_result(|| {
+    ///         let _v = may_fail_with_placeholder_error()?;
+    ///
+    ///         Ok(())
+    ///     })
+    /// }
+    ///
+    /// fn may_fail_with_placeholder_error() -> Result<u8, PlaceholderError> {
+    ///     Ok(42)
+    /// }
+    /// ```
+    pub fn capture_into_result<T>(body: impl FnOnce() -> Result<T, E>) -> Result<T, Self> {
+        body().map_err(Self::from_error)
+    }
+
+    /// Executes a closure that returns a [`Result`][], converting any
+    /// error to a [`Report`][].
+    ///
+    /// Recommended if you only support Rust version 1.61 or above.
+    ///
+    /// ```rust
+    /// use snafu::{prelude::*, Report};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// struct PlaceholderError;
+    ///
+    /// fn main() -> Report<PlaceholderError> {
+    ///     Report::capture(|| {
+    ///         let _v = may_fail_with_placeholder_error()?;
+    ///
+    ///         Ok(())
+    ///     })
+    /// }
```

Comment:
This doctest should only run this `main` function when supporting Rust 1.61+. The test passes with default Cargo features after this change.
```suggestion
    /// # #[cfg(feature = "rust_1_61")]
    /// fn main() -> Report<PlaceholderError> {
    ///     Report::capture(|| {
    ///         let _v = may_fail_with_placeholder_error()?;
    ///
    ///         Ok(())
    ///     })
    /// }
    /// # #[cfg(not(feature = "rust_1_61"))]
    /// #  fn main() {}
```

Reply:
Yep. Dang tests.